### PR TITLE
CSHARP-4292: Performance: Eliminated allocations in GetHashCode methods

### PR DESCRIPTION
**Pull request by dnickless**
Eliminated some boxing problems leading to substantial allocations in our code. Details located here: https://jira.mongodb.org/browse/CSHARP-4292

### CODE REVIEW

File: src/MongoDB.Shared/Hasher.cs [hunk 1]
```diff
@@ -75,17 +75,17 @@ public Hasher Hash(object obj)
...
             {
                 _hashCode = 37 * _hashCode + -1;
             }
             else
             {
-                foreach (var value in sequence)
+                foreach (T value in sequence)
                 {
-                    _hashCode = 37 * _hashCode + ((value == null) ? -1 : value.GetHashCode());
+                    _hashCode = 37 * _hashCode + (value == null ? -1 : EqualityComparer<T>.Default.GetHashCode());
```
Comment (BorisDog) [inline src/MongoDB.Shared/Hasher.cs:88]: Could you please wrap `EqualityComparer<T>.Default.GetHashCode())` in

 ```
internal class Hasher
 {
      #region static 

      // public static methods
      [MethodImpl(MethodImplOptions.AggressiveInlining)]
      public static int GetHashCode<T>(T obj) => EqualityComparer<T>.Default.GetHashCode(obj);

      #end

      // private fields
      private int _hashCode;
      ...
 }
```

**Review by BorisDog — CHANGES_REQUESTED** [review]
Good optimization overall. Minor comments.

File: src/MongoDB.Shared/Hasher.cs (hunk 1 above)
Comment (BorisDog) [inline src/MongoDB.Shared/Hasher.cs:88]: missing `value` in `EqualityComparer<T>.Default.GetHashCode(**value**).` 
This was caught by local tests.

File: src/MongoDB.Bson/ObjectModel/BsonRegularExpression.cs [hunk 2]
```diff
@@ -243,9 +244,9 @@ public override int GetHashCode()
         {
             // see Effective Java by Joshua Bloch
             int hash = 17;
-            hash = 37 * hash + BsonType.GetHashCode();
-            hash = 37 * hash + _pattern.GetHashCode();
-            hash = 37 * hash + _options.GetHashCode();
+            hash = 37 * hash + EqualityComparer<BsonType>.Default.GetHashCode(BsonType);
+            hash = 37 * hash + EqualityComparer<string>.Default.GetHashCode(_pattern);
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Bson/ObjectModel/BsonRegularExpression.cs:248]: I don't think it's necessary for value types either, just for enums.

**Comment by dnickless** [comment]
Everything fixed and cleaned up now using separate commits for easier change tracking - should arguably get squashed.

File: src/MongoDB.Bson/ObjectModel/BsonBinaryData.cs [hunk 3]
```diff
@@ -349,12 +350,12 @@ public override int GetHashCode()
             // see Effective Java by Joshua Bloch
             // note: guidRepresentation is not considered when computing the hash code
             int hash = 17;
-            hash = 37 * hash + BsonType.GetHashCode();
+            hash = 37 * hash + EqualityComparer<BsonType>.Default.GetHashCode(BsonType);
```
Comment (BorisDog) [inline src/MongoDB.Bson/ObjectModel/BsonBinaryData.cs:353]: Could you please substitute all `EqualityComparer<T>.Default` with `Hasher.GetHashCode`?

File: src/MongoDB.Bson/ObjectModel/BsonRegularExpression.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Bson/ObjectModel/BsonRegularExpression.cs]: Good point @rstam.

**Review by BorisDog — CHANGES_REQUESTED** [review]
Thanks, one last comment.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**